utils/cache/filecache/filecache_test.go

```diff
@@ -0,0 +1,754 @@
+			require.NoError(t, err)
+
+			slices.Sort(srcRelTree)
+			slices.Sort(cacheRelTree)
+			require.Equal(t, srcRelTree, cacheRelTree, "Cache Dir and Src Dir have differnet contents")
+		})
+	}
+
+	t.Run("Add_invalid_path", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		fs := filesystem.NewFs(filesystem.StandardFS)
+		tmpCacheDir, err := fs.TempDirInTempDir("test-cache")
+		require.NoError(t, err)
+		defer func() { _ = fs.Rm(tmpCacheDir) }()
+
+		ctx := context.Background()
+		config := DefaultFileCacheConfig()
+		config.CachePath = tmpCacheDir
+		cache, err := NewFsFileCache(ctx, fs, fs, "/does/not/exist", config)
+		require.NoError(t, err)
+
+		defer func() {
+			err = cache.Close(ctx)
+			require.NoError(t, err)
+		}()
+
+		err = cache.Store(ctx, faker.UUIDDigit())
+		errortest.AssertError(t, err, commonerrors.ErrNotFound)
+	})
+
+	t.Run("Add_duplicate_entry", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		fs := filesystem.NewFs(filesystem.StandardFS)
+		tmpCacheDir, err := fs.TempDirInTempDir("test-cache")
+		require.NoError(t, err)
+		defer func() { _ = fs.Rm(tmpCacheDir) }()
+
+		tmpTestPath, err := fs.TouchTempFileInTempDir(faker.Word())
+		require.NoError(t, err)
+		tmpTestDir := fs.TempDirectory()
+		tmpTestBase := filesystem.FilePathBase(fs, tmpTestPath)
+
+		ctx := context.Background()
+		config := DefaultFileCacheConfig()
+		config.CachePath = tmpCacheDir
+		cache, err := NewFsFileCache(ctx, fs, fs, tmpTestDir, config)
+		require.NoError(t, err)
+
+		defer func() {
+			err = cache.Close(ctx)
+			require.NoError(t, err)
+		}()
+
+		err = cache.Store(ctx, tmpTestBase)
+		require.NoError(t, err)
+
+		err = cache.Store(ctx, tmpTestBase)
+		errortest.AssertError(t, err, commonerrors.ErrExists)
+	})
+}
+
+func TestFileCache_Fetch(t *testing.T) {
+	t.Run("Fetch_file", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		fs := filesystem.NewFs(filesystem.StandardFS)
+		tmpCacheDir, err := fs.TempDirInTempDir("test-cache")
+		require.NoError(t, err)
+		defer func() { _ = fs.Rm(tmpCacheDir) }()
+
+		tmpfilePath, err := fs.TouchTempFileInTempDir(faker.Word())
+		require.NoError(t, err)
+		tmpfileDir := fs.TempDirectory()
+		tmpfileBase := filesystem.FilePathBase(fs, tmpfilePath)
+
+		ctx := context.Background()
+		config := DefaultFileCacheConfig()
+		config.CachePath = tmpCacheDir
+		cache, err := NewFsFileCache(ctx, fs, fs, tmpfileDir, config)
+		require.NoError(t, err)
+
+		defer func() {
+			err = cache.Close(ctx)
+			require.NoError(t, err)
+		}()
+
+		err = cache.Store(ctx, tmpfileBase)
+		require.NoError(t, err)
+
+		err = fs.Rm(tmpfilePath)
+		require.NoError(t, err)
+		require.False(t, filesystem.Exists(tmpfilePath), "could not remove the src file")
+
+		err = cache.Fetch(ctx, tmpfileBase, fs, tmpfilePath)
+		require.NoError(t, err)
+
+		require.True(t, filesystem.Exists(tmpfilePath), "cache did not fetch the file")
+	})
+
+	t.Run("Fetch_file_non_existent", func(t *testing.T) {
```

Comment:
Just so you're aware, you should make the test names human readable in the source code. 

When we run the tests it is go that will handle the underscores and stuff.